File: src/examples/findFirst/findFirst.cpp
/*-------------------------------------------------------------------------------------------
 * qblocks - fast, easily-accessible, fully-decentralized data from blockchains
 * copyright (c) 2016, 2021 TrueBlocks, LLC (http://trueblocks.io)
 *
 * This program is free software: you may redistribute it and/or modify it under the terms
 * of the GNU General Public License as published by the Free Software Foundation, either
 * version 3 of the License, or (at your option) any later version. This program is
 * distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License for more details. You should have received a copy of the GNU General
 * Public License along with this program. If not, see http://www.gnu.org/licenses/.
 *-------------------------------------------------------------------------------------------*/
#include "etherlib.h"

//-----------------------------------------------------------------------------------------------
// The following code shows an example of using one of TrueBlocks' forEvery features. forEvery
// does as it sounds. It visits every item in a given data structure. Here, we're showing
// an example of showing every address appearance in a block. We're searching for the first
// transaction in which the infamous The DAO (0xbb9...) appears. There are many forEvery functions
// such as forEveryTransactionInBlock, forEveryTraceInTransaction, forEveryLogInBlock, etc.

//-----------------------------------------------------------------------------------------------
// Every forEvery invocation accepts a void* to a chunk of memory that can be of any type
// you wish and a function. The function signature of that function differs for each forEvery
// type. In this case, the signature must be as follows.
bool visitAddress(const CAppearance& item, void* data) {
    if (item.addr == *reinterpret_cast<address_t*>(data)) {
        cout << "Found at " << item << "\n";
        return false;  // return false to stop the scan
    }
    cerr << item << "                    \r";
    return true;  // return true to continue
}

//-----------------------------------------------------------------------------------------------
int main(int argc, const char* argv[]) {
    // Initialize the library
    etherlib_init(quickQuitHandler);

    // a best guess place to start the scan
    blknum_t startBlock = 1428000;

    // the address we're searching for
    address_t theDaoAddr("0xbb9bc244d798123fde783fcc1c72d3bb8c189413");

    // scan each block from the start until we find the address in an appearance
    for (blknum_t bl = startBlock; bl < getBlockProgress(BP_CLIENT).client; bl++) {
        CBlock block;
        getBlock(block, bl);
        if (!block.forEveryUniqueAppearanceInBlock(visitAddress /* func */, NULL /* filterFunc */,
                                                   &theDaoAddr /* data */))
            return 0;
    }

    etherlib_cleanup();

    return 0;
}

File: src/examples/findFirst/findFirst.go
package main

import (
	"context"
	"fmt"
	"time"

	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/base"
	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/identifiers"
	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/logger"
	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/rpc"
	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/types"
	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/utils"
	"github.com/TrueBlocks/trueblocks-core/src/apps/chifra/pkg/validate"
)

// -----------------------------------------------------------------------------------------------
// The following code shows an example of using one of TrueBlocks' forEvery features. forEvery
// does as it sounds. It visits every item in a given data structure performing a particular
// function. Here, we're showing an example of showing every address appearance in a block.
// We're searching for the first transaction in which the infamous The DAO (0xbb9...)
// appears. There are many forEvery functions such as forEveryTransaction,
// forEveryTrace, forEveryLog, etc.
// -----------------------------------------------------------------------------------------------
func main() {
	chain := "mainnet"
	conn := rpc.TempConnection(chain)
	slowWay(conn)
	fastWay(conn)
}

var chain = "mainnet"

func slowWay(conn *rpc.Connection) {
	start := time.Now()
	bar := logger.NewBar(logger.BarOptions{
		Prefix:  "Getting stuff",
		Enabled: true,
		Start:   40000,
		Total:   60000,
	})
	for i := 40000; i < 60000; i++ {
		if block, err := conn.GetBlockHeaderByNumber(base.Blknum(i)); err != nil {
			fmt.Println(err)
		} else {
			if len(block.Transactions) > 0 {
				bar.Finish(true /* newLine */)
				done := time.Since(start)
				fmt.Println("Time taken:", done, "seconds")
				fmt.Println(i, block.Hash.String(), "in", done, "seconds")
				break
			} else {
				bar.Tick()
			}
		}
	}
}

func fastWay(conn *rpc.Connection) {
	var BlockIds []identifiers.Identifier
	if err := validate.ValidateIdentifiers(chain, []string{"40000-60000"}, validate.ValidBlockIdWithRange, 1, &BlockIds); err != nil {
		fmt.Println(err)
	}

	apps, _, err := identifiers.IdsToApps(chain, BlockIds)
	if err != nil {
		fmt.Println(err)
	}

	if sliceOfMaps, cnt, err := types.AsSliceOfMaps[types.Block[string]](apps, false); err != nil {
		fmt.Println(err)

	} else if cnt != 0 {
		fmt.Println("Expected a block")

	} else {
		bar := logger.NewBar(logger.BarOptions{
			Type:    logger.Fixed,
			Prefix:  "Getting stuff",
			Enabled: true,
			Start:   40000,
			Total:   60000,
		})

		for _, thisMap := range sliceOfMaps {
			for app := range thisMap {
				thisMap[app] = new(types.Block[string])
			}

			var firstBlock types.Block[string]
			firstBlock.BlockNumber = utils.NOPOS
			iterateFunc := func(key types.Appearance, value *types.Block[string]) error {
				bn := uint64(key.BlockNumber)
				if theBlock, err := conn.GetBlockHeaderByNumber(base.Blknum(key.BlockNumber)); err != nil {
					return err
				} else {
					if len(theBlock.Transactions) > 0 {
						// fmt.Println("Found", theBlock.Hash.String(), ":", len(theBlock.Transactions), "transactions")
						if theBlock.BlockNumber > 0 && bn < firstBlock.BlockNumber {
							firstBlock = theBlock
							fmt.Println("    Set", firstBlock.BlockNumber, firstBlock.Hash.String(), ":", len(firstBlock.Transactions), "transactions             ")
						}
					}
					bar.Tick()
					return nil
				}
			}

			errorChan := make(chan error)
			ctx, cancel := context.WithCancel(context.Background())
			defer cancel()
			go utils.IterateOverMap(ctx, errorChan, thisMap, iterateFunc)
			for err := range errorChan {
				fmt.Println(err)
			}

			timeTaken := bar.Finish(true /* newLine */)
			fmt.Println(firstBlock.BlockNumber, firstBlock.Hash.String(), "in", timeTaken, "seconds")
		}
		bar.Finish(true /* newLine */)
	}
}
